ApplicationLibCode/ProjectDataModel/RimUserDefinedCalculation.cpp

```diff
@@ -0,0 +1,382 @@
+void RimUserDefinedCalculation::fieldChangedByUi( const caf::PdmFieldHandle* changedField,
+                                                  const QVariant&            oldValue,
+                                                  const QVariant&            newValue )
+{
+    m_isDirty = true;
+
+    PdmObject::fieldChangedByUi( changedField, oldValue, newValue );
+}
+
+//--------------------------------------------------------------------------------------------------
+///
+//--------------------------------------------------------------------------------------------------
+RimUserDefinedCalculationVariable* RimUserDefinedCalculation::findByName( const QString& name ) const
+{
+    for ( RimUserDefinedCalculationVariable* v : m_variables )
+    {
+        if ( v->name() == name )
+        {
+            return v;
+        }
+    }
+
+    return nullptr;
+}
+
+//--------------------------------------------------------------------------------------------------
+///
+//--------------------------------------------------------------------------------------------------
+QString RimUserDefinedCalculation::buildCalculationName() const
+{
+    QString name = "Default Calculation Name";
+
+    QString lhs = RimUserDefinedCalculation::findLeftHandSide( m_expression );
+    if ( !lhs.isEmpty() )
+    {
+        name = lhs;
+
+        name += " ( ";
+
+        for ( RimUserDefinedCalculationVariable* v : m_variables )
+        {
+            name += v->displayString();
+
+            if ( v != m_variables[m_variables.size() - 1] )
+            {
+                name += ", ";
+            }
+        }
+
+        name += " )";
+    }
+
+    return name;
+}
+
+//--------------------------------------------------------------------------------------------------
+///
+//--------------------------------------------------------------------------------------------------
+void RimUserDefinedCalculation::defineEditorAttribute( const caf::PdmFieldHandle* field,
+                                                       QString                    uiConfigName,
+                                                       caf::PdmUiEditorAttribute* attribute )
+{
+    if ( field == &m_expression )
+    {
+        caf::PdmUiTextEditorAttribute* myAttr = dynamic_cast<caf::PdmUiTextEditorAttribute*>( attribute );
+        if ( myAttr )
+        {
+            myAttr->heightHint = -1;
+        }
+    }
+    else if ( field == &m_variables )
+    {
+        auto* myAttr = dynamic_cast<caf::PdmUiTableViewEditorAttribute*>( attribute );
+        if ( myAttr )
+        {
+            myAttr->enableDropTarget = true;
+        }
+    }
+}
+
+//--------------------------------------------------------------------------------------------------
+///
+//--------------------------------------------------------------------------------------------------
+std::vector<RimUserDefinedCalculationVariable*> RimUserDefinedCalculation::allVariables() const
```

Comment:
To access all objects as a std::vector you can use `caf::PdmChildArrayField<DataType*>::childObjects()`

Reply:
Good catch!